File: leetcode/22. Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(res, "", 0, 0, n);
        return res;
    }
    
private:
    void helper(vector<string>& res, string str, int left, int right, int n){
        if(left == n){
            for(int i = right; i < left; ++i){
                str += ')';
            }
            res.push_back(str);
            return;
        }
        
        helper(res, str + '(', left + 1, right, n);
        if(right < left){
            helper(res, str + ')', left, right + 1, n);
        }
    }
};

File: leetcode/backtrack/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def helper(left, right, curr):
            if left == n and right == n:
                res.append(''.join(curr))
                return
            
            if left < n:
                curr.append('(')
                helper(left+1, right, curr)
                curr.pop()
            
            if right < left:
                curr.append(')')
                helper(left, right+1, curr)
                curr.pop()
            
        helper(0, 0, [])
        return res
